source/brailleDisplayDrivers/hidBrailleStandard.py

```diff
@@ -103,16 +104,27 @@ def __init__(self, port="auto"):
 				continue  # Couldn't connect.
 			if self._dev.usagePage != HID_USAGE_PAGE_BRAILLE:
 				log.debug("Not braille")
+				self._dev.close()
 				continue
 			cellValueCaps = self._findCellValueCaps()
-			if cellValueCaps:
+			if len(cellValueCaps) > 0:
+				if any(x.ReportCount != cellValueCaps[0].ReportCount for x in cellValueCaps):
+					log.warn("Found multi-line display with an irregular shape, ignoring.")
+					self._dev.close()
+					continue
+				self.numRows = len(cellValueCaps)
+				self.numCols = cellValueCaps[0].ReportCount
+				self._maxNumberOfCells = self.numCells
 				self._cellValueCaps = cellValueCaps
-				self._numberOfCellsValueCaps = self._findNumberOfCellsValueCaps()
-				self.numCells = self._maxNumberOfCells = cellValueCaps.ReportCount
+				if self.numRows == 1:
+					self._numberOfCellsValueCaps = self._findNumberOfCellsValueCaps()
+				elif self._findNumberOfCellsValueCaps():
+					log.warn("The number of braille cells usage is not supported on multi-line displays")
```

Comment:
I don't quite understand this message. Can you explain this case in more detail so I can suggest an alternative wording?

Reply:
The HID standard defines a usage for "Number of Braille Cells" which is described as follows:

> Some braille displays dynamically reserve a portion of a braille row for display specific behavior, for example showing the progress of a file transfer. For example, a 20 cell display might reserve 4 cells. If this usage was set to 16, then cells 17 through 20 would be ignored by the braille display

This usage is applied to a braille row and applies to the cells in that row. On a single-line display this is fairly straightforward to deal with as changes to that value changes the effective number of cells of the display. However, on a multi-line display individual rows changing their effective length would likely lead to an irregular grid of cells which the BrailleDisplay interface doesn't have a way to represent. I'm not aware of any multi-line braille displays that are using that usage in their descriptors, so I think it makes the most sense to just ignore it and log a warning in case a future braille display does attempt to use that capability.